runtime/src/accounts_db.rs

```diff
@@ -1559,47 +1559,52 @@ impl AccountsDb {
+                                                    &locked_entry,
+                                                    max_clean_root,
+                                                ),
+                                            );
+                                        }
+                                        // prune zero_lamport_pubkey set which should contain all 0-lamport
+                                        // keys whether rooted or not. A 0-lamport update may become rooted
+                                        // in the future.
+                                        let remove_from_zero_lamports_list =
+                                            !slot_list.iter().any(|(_slot, account_info)| {
+                                                account_info.lamports == 0
+                                            });
+                                        // Release the lock
+                                        let slot = *slot;
+                                        drop(locked_entry);
+
+                                        if self.accounts_index.is_uncleaned_root(slot) {
+                                            // Assertion enforced by `accounts_index.get()`, the latest slot
+                                            // will not be greater than the given `max_clean_root`
+                                            if let Some(max_clean_root) = max_clean_root {
+                                                assert!(slot <= max_clean_root);
+                                            }
+                                            purges_in_root.push(*pubkey);
+                                        }
+                                        remove_from_zero_lamports_list
                                     }
-                                    purges_in_root.push(*pubkey);
-                                }
-                            } else {
-                                let r_accounts_index =
-                                    self.accounts_index.account_maps.read().unwrap();
-                                if !r_accounts_index.contains_key(pubkey) {
-                                    self.accounts_index.remove_zero_lamport_key(pubkey);
-                                }
+                                    AccountIndexGetResult::NoRootFound(_locked_entry) => {
+                                        // do nothing - pubkey is in index, but not found in a root slot
+                                        false
+                                    }
+                                    AccountIndexGetResult::Missing() => {
+                                        // pubkey is missing from index, so remove from zero_lamports_list
+                                        true
+                                    }
+                                };
+                            if remove_from_zero_lamports_list {
```

Comment:
hmmm I think the issue here is the lock is not held for this pubkey during the remove.  This opens the window for the pubkey to be reinserted by some ongoing slot processing  between when we checked for emptiness, dropped the lock, and then delete here

Reply:
Hmmm in the original code there are two calls to `remove_from_zero_lamports_list`:
1) The first https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1577-L1583 is safe because the slot list lock for that particular pubkey is being held, so it cannot be modified between the check: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1578-L1580 and the subsequent remove.

2) This path here https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1598-L1602 holds the `account_maps` lock between the check and delete

This new change does not hold the lock through either of these checks anymore, i.e., 

1) the slot list lock is dropped here: https://github.com/solana-labs/solana/pull/16689/files#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR1585 after checking for any zero lamport accounts, so that check can be outdated by the time the delete is called
2) If the `NotFoundOnFork` is triggered, we don't grab the lock again as described in 2) above in the original code to check for any changes, so this check can be outdated by the time the delete is called

Reply:
yes, I agree with your conclusions and arrived at them as well. I should have taken the time to be more precise and complete in my answer earlier. I was demonstrating that there are 2 different locks that are being held for the different code paths. I haven't done the work yet to figure out the other writers of this list and under what conditions they would be allowed to write to it with no synchronization enforced on it. I meant to do that work before you had read my earlier comment, but I was distracted all day so far. I was not suggesting that the code as it is in this pr is thus correct - merely trying to understand and make sure that the old and new code path are well understood and correct. You raised a very good point initially that there could be synchronization problems on changes to the zero key list code. It is obvious there are subtle synchronization issues here that I don't understand yet. I can fix this by doing the list removes within the lock. I will also figure out how synchronization and access to this list actually works.

Reply:
Ah. zero_lamport_pubkeys is a dashset. And modifications to it are designed to be governed indirectly by one of (at least) 2 different external (external to the dashset) locks.

Reply:
🦶 💣